File: packages/ruby-sdk/lib/strait/composition/paginate.rb
# frozen_string_literal: true

module Strait
  module Composition
    PaginatedQuery = Struct.new(:cursor, :limit, keyword_init: true)

    PaginatedResponse = Struct.new(:data, :items, :next_cursor, :has_more, keyword_init: true)

    PaginateOptions = Struct.new(:limit, keyword_init: true) do
      def initialize(**kwargs)
        super
        self.limit ||= 50
      end
    end

    def self.paginate(opts = nil, &list_fn)
      opts ||= PaginateOptions.new
      Enumerator.new do |yielder|
        cursor = nil
        loop do
          query = PaginatedQuery.new(cursor: cursor, limit: opts.limit)
          response = list_fn.call(query)
          items = response.items || response.data || []
          items.each { |item| yielder << item }

          break unless response.has_more
          break if response.next_cursor.nil? || response.next_cursor.empty?
          cursor = response.next_cursor
        end
      end
    end

    def self.collect_all(opts = nil, &list_fn)
      paginate(opts, &list_fn).to_a
    end
  end
end

File: packages/rust-sdk/src/composition/paginate.rs
use crate::errors::StraitError;
use std::future::Future;

#[derive(Debug, Clone, Default)]
pub struct PaginatedQuery {
    pub cursor: Option<String>,
    pub limit: Option<u32>,
}

#[derive(Debug, Clone)]
pub struct PaginatedResponse<T> {
    pub data: Vec<T>,
    pub next_cursor: Option<String>,
    pub has_more: bool,
}

#[derive(Debug, Clone)]
pub struct PaginateOptions {
    pub limit: u32,
}

impl Default for PaginateOptions {
    fn default() -> Self {
        Self { limit: 50 }
    }
}

pub async fn collect_all<T, F, Fut>(
    list_fn: F,
    opts: Option<&PaginateOptions>,
) -> Result<Vec<T>, StraitError>
where
    F: Fn(PaginatedQuery) -> Fut,
    Fut: Future<Output = Result<PaginatedResponse<T>, StraitError>>,
{
    let default_opts = PaginateOptions::default();
    let opts = opts.unwrap_or(&default_opts);
    let mut all_items = Vec::new();
    let mut cursor = None;

    loop {
        let query = PaginatedQuery {
            cursor: cursor.clone(),
            limit: Some(opts.limit),
        };
        let response = list_fn(query).await?;
        all_items.extend(response.data);

        if !response.has_more || response.next_cursor.is_none() {
            break;
        }
        cursor = response.next_cursor;
    }

    Ok(all_items)
}
